crates/sage/src/enzyme.rs

```diff
@@ -141,75 +161,135 @@ impl Enzyme {
             None => {
                 // Perform a non-specific digest
                 let mut v = Vec::new();
                 for len in self.min_len..=self.max_len {
                     for i in 0..=sequence.len().saturating_sub(len) {
-                        v.push(i..i + len)
+                        v.push(DigestSite{
+                            site: i..i + len,
+                            missed_cleavages: 0}
+                        );
                     }
                 }
                 v
             }
         }
     }
 
+    fn missed_cleavage_sites(&self, sites: &mut Vec<DigestSite>, missed_cleavages: &u8) -> Vec<DigestSite> {
+        let mut missed_cleavage_sites = Vec::new();
+        for cleavage in 1..=(1 + missed_cleavages) {
+            // Generate missed cleavages
+            for win in sites.windows(cleavage as usize) {
+                let start = win[0].site.start;
+                let end = win[cleavage as usize - 1].site.end;
+                missed_cleavage_sites.push(DigestSite{
+                    site: start..end,
+                    missed_cleavages: cleavage - 1,
+                });
+            }
+        }
+        sites.append(&mut missed_cleavage_sites);
+        sites.to_vec()
+    }
+
+    fn is_semi_enzymatic(&self) -> bool {
+        match &self.enyzme {
+            Some(enzyme) => enzyme.is_semi_enzymatic,
+            None => false,
+        }
+    }
+
+    fn semi_enzymatic_sites(&self, sites: &mut Vec<DigestSite>) -> Vec<DigestSite> {
+        let mut semi_enzymatic_sites = Vec::new();
+        for site in sites.iter_mut() {
+            let start = site.site.start;
+            let end = site.site.end;
+            for cut_site in start..end {
+                semi_enzymatic_sites.push(DigestSite{
+                    site: start..cut_site,
+                    missed_cleavages: site.missed_cleavages
+                });
+                semi_enzymatic_sites.push(DigestSite{
+                    site: cut_site..end, 
+                    missed_cleavages: site.missed_cleavages
+                });
+            }
+        }
+        sites.append(&mut semi_enzymatic_sites);
+        sites.to_vec()
+    }
+
     pub fn digest(&self, sequence: &str, protein: Arc<String>) -> Vec<Digest> {
         let n = sequence.len();
         let mut digests = Vec::new();
-        let sites = self.cleavage_sites(sequence);
+        let mut sites = self.cleavage_sites(sequence);
         // Allowing missed_cleavages with non-specific digest causes OOB panics
         // in the below indexing code
         let missed_cleavages = match self.enyzme {
             None => 0,
             _ => self.missed_cleavages,
         };
 
+        let mut sites = match missed_cleavages{
+            0 => sites,
+            _ => self.missed_cleavage_sites(&mut sites, &missed_cleavages)
+        };
+
+        let mut sites = match self.is_semi_enzymatic() {
```

Comment:
I keep getting `error[E0609]: no field `enzyme` on type `&EnzymeParameters`` at compile time when I match to `self.enzyme.is_semi_enzymatic`. That was my first instinct until I kept getting the error.

Reply:
Ah, it's because it's an `Option` - will need to be something like `self.enzyme.map(|e| e.is_semi_enzymatic).unwrap_or(false)`... which is not as concise as I initially imagined. We can just leave it as-is for now